### PR TITLE
[ConstraintElim] Add facts implied by llvm.abs

### PR DESCRIPTION
**Pull request by alexander-shaposhnikov**
Add  "abs(x) >=s x" fact.

https://alive2.llvm.org/ce/z/gOrrU3

Test plan: ninja check-all
Happy to add more tests covering llvm.abs.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
What is the motivation of this patch? I think it is easy to implement this fold in `InstSimplify`.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 1]
```diff
@@ -995,6 +995,11 @@ void State::addInfoFor(BasicBlock &BB) {
       continue;
     }
 
+    if (match(&I, m_Intrinsic<Intrinsic::abs>())) {
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:998]: Now that we support a number of intrinsics, we should probably covert that to a switch on the intrinsic ID, so it is easy to see what intrinsics are handled at a glance, as well as making it obvious that they are all handled in the same way

**Review by fhahn — COMMENTED** [review]
Thanks for the patch! IMO it is very worthwhile to teach ConstraintElim about more intrinsics, as this can enable additional optimizations by the pass that cannot easily be done by other passes. 

@Alexander Shaposhnikov Assuming you still have the original motivating case, any chance you could add a more complicated test case that can only be handled by `ConstraintElim`?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 2]
```diff
@@ -984,32 +984,38 @@ void State::addInfoFor(BasicBlock &BB) {
...
     }
 
-    if (match(&I, m_Intrinsic<Intrinsic::ssub_with_overflow>())) {
+    auto *II = dyn_cast<IntrinsicInst>(&I);
+    Intrinsic::ID ID = II ? II->getIntrinsicID() : Intrinsic::not_intrinsic;
+    switch (ID) {
+    case Intrinsic::assume:
+      Value *A, *B;
+      CmpInst::Predicate Pred;
+      if (I.getNumOperands() >= 1 &&
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:993]: Is this needed? `assume` calls should only have a single argument?

File: llvm/test/Transforms/ConstraintElimination/abs.ll [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+define i1 @abs_plus_one(i32 %arg) {
+; CHECK-LABEL: define i1 @abs_plus_one(
+; CHECK-SAME: i32 [[ARG:%.*]]) {
+; CHECK-NEXT:    [[ABS:%.*]] = tail call i32 @llvm.abs.i32(i32 [[ARG]], i1 true)
+; CHECK-NEXT:    [[ABS_PLUS_ONE:%.*]] = add nsw i32 [[ABS]], 1
+; CHECK-NEXT:    ret i1 true
+;
+  %abs = tail call i32 @llvm.abs.i32(i32 %arg, i1 true)
```
Comment (fhahn) [inline llvm/test/Transforms/ConstraintElimination/abs.ll:33]: To clarify, I think that's the extra test, right?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 2 above)
Comment (alexander-shaposhnikov) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:993]: you are right,  not needed

File: llvm/test/Transforms/ConstraintElimination/abs.ll (hunk 3 above)
Comment (alexander-shaposhnikov) [reply llvm/test/Transforms/ConstraintElimination/abs.ll:33]: yup

**Comment by alexander-shaposhnikov** [comment]
Address comments

File: llvm/test/Transforms/ConstraintElimination/abs.ll [hunk 4]
```diff
@@ -0,0 +1,39 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -passes=constraint-elimination -S %s | FileCheck %s
+
+define i1 @abs_int_min_is_not_poison(i32 %arg) {
```
Comment (fhahn) [inline llvm/test/Transforms/ConstraintElimination/abs.ll:4]: Could you submit the tests separately and also add a version with constant arguments (both positive and negative)?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 5]
```diff
@@ -1629,6 +1634,11 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT, LoopInfo &LI,
 
     ICmpInst::Predicate Pred;
     if (!CB.isConditionFact()) {
+      if (Value *X; match(CB.Inst, m_Intrinsic<Intrinsic::abs>(m_Value(X)))) {
+        AddFact(CmpInst::ICMP_SGE, CB.Inst, X);
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1638]: I think we should also be able to add `CB.Inst >= 0` as a follow-up?

File: llvm/test/Transforms/ConstraintElimination/abs.ll (hunk 4 above)
Comment (alexander-shaposhnikov) [reply llvm/test/Transforms/ConstraintElimination/abs.ll:4]: ping @fhahn - happy to send a separate PR with the tests, but need a clarification regarding "constant arguments" - what did you mean

Comment (fhahn) [reply llvm/test/Transforms/ConstraintElimination/abs.ll:4]: Constants to the llvm.abs calls

**Comment by alexander-shaposhnikov** [comment]
@fhahn - sorry about the delay, just sent a separate PR with tests:
https://github.com/llvm/llvm-project/pull/76374

**Comment by dtcxzyw** [comment]
What is the motivation of this patch? Is this pattern from a real-world application? IMHO we should not add optimizations to improve the "coverage".

**Comment by alexander-shaposhnikov** [comment]
@dtcxzyw  - yes, llvm.abs calls originate e.g. from std::abs and this showed up multiple times when i was testing the patch internally.

**Comment by dtcxzyw** [comment]
> @dtcxzyw - yes, llvm.abs calls originate e.g. from std::abs and this showed up multiple times when i was testing the patch internally.

Great! Please add a more complicated test that can only be handled by `ConstraintElim` as @fhahn suggested.
Could you please provide the link to the code repo? I would like to add it into my benchmark :)

**Comment by alexander-shaposhnikov** [comment]
@dtcxzyw , many thanks, actually the suggested test has already been added (GitHub seems to hide the resolved comment threads) (see https://github.com/llvm/llvm-project/pull/73189/commits/c13819ccea048833d24214221cc52d0b52eaf4a9)

**Comment by dtcxzyw** [comment]
> @dtcxzyw , many thanks, actually the suggested test has already been added (GitHub seems to hide resolved comment threads) (see [c13819c](https://github.com/llvm/llvm-project/commit/c13819ccea048833d24214221cc52d0b52eaf4a9))

All the test cases should be handled by `InstSimplify`.

> Could you please provide the link to the code repo?

I mean the link to the `real-world application".

**Comment by alexander-shaposhnikov** [comment]
@dtcxzyw

> All the test cases should be handled by InstSimplify

The test that involves arithmetic was intentionally made simple. Regarding InstSimplify - unfortunately I would probably disagree with the approach. ConstraintElim is a more general / robust solution and iirc it was introduced for these purposes (as an alternative to expanding the complexity of patterns in InstCombine/InstSimplify), but I'd allow @nikic and @fhahn to weigh in / correct me. I think I've seen a few more cases where  ConstraintElim stumbles on intrinsics, this was just the simplest one.

> Could you please provide the link to the code repo?
The codebase is internal unfortunately.
I've just checked the list of github issues - it looks like it was independently noticed by other people here https://github.com/llvm/llvm-project/issues/72653

**Comment by dtcxzyw** [comment]
`abs(x) <=u x` also holds.
Alive2: https://alive2.llvm.org/ce/z/x4fPYp

**Comment by alexander-shaposhnikov** [comment]
P.S. I'd like to expand my previous comment a bit - I've just checked and currently InstSimplify doesn't handle these cases
(if the args are not constant):
```
./bin/opt -passes=instsimplify -S ~/abs.ll 
```
```
define i1 @abs_int_min_is_not_poison(i32 %arg) {
  %abs = tail call i32 @llvm.abs.i32(i32 %arg, i1 false)
  %cmp = icmp sge i32 %abs, %arg
  ret i1 %cmp
}

define i1 @abs_int_min_is_poison(i32 %arg) {
  %abs = tail call i32 @llvm.abs.i32(i32 %arg, i1 true)
  %cmp = icmp sge i32 %abs, %arg
  ret i1 %cmp
}

define i1 @abs_plus_one(i32 %arg) {
  %abs = tail call i32 @llvm.abs.i32(i32 %arg, i1 true)
  %abs_plus_one = add nsw i32 %abs, 1
  %cmp = icmp sge i32 %abs_plus_one, %arg
  ret i1 %cmp
}

define i1 @abs_constant_negative_arg() {
  ret i1 true
}

define i1 @abs_constant_positive_arg() {
  ret i1 true
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare i32 @llvm.abs.i32(i32, i1 immarg) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```
To me teaching ConstrainElim about intrinsics seems to be a better idea than putting more patterns into InstSimplify,
but maybe I'm missing something.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 5 above)
Comment (fhahn) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1638]: Sounds good, could you add a TODO?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 6]
```diff
@@ -1629,6 +1634,11 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT, LoopInfo &LI,
...
+        continue;
+      }
+
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1641]: I think all non-conditional facts should be handled here, so could we add `llvm_unreachable()` below and remove the extra check `    if (CB.isConditionFact()) {` as a follow-up?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 7]
```diff
@@ -984,32 +984,37 @@ void State::addInfoFor(BasicBlock &BB) {
...
+          WorkList.emplace_back(FactOrCheck::getConditionFact(
+              DT.getNode(I.getParent()), Pred, A, B));
+        } else {
+          WorkList.emplace_back(
+              FactOrCheck::getInstFact(DT.getNode(I.getParent()), &I));
+        }
+      }
+      break;
+    case Intrinsic::ssub_with_overflow:
       WorkList.push_back(
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1006]: might be good to add a comment here that we queue ssub_with_overflow for simplification and below we queue the intrinsics to add extra info.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 8]
```diff
@@ -984,32 +984,37 @@ void State::addInfoFor(BasicBlock &BB) {
       continue;
     }
 
-    if (match(&I, m_Intrinsic<Intrinsic::ssub_with_overflow>())) {
+    auto *II = dyn_cast<IntrinsicInst>(&I);
+    Intrinsic::ID ID = II ? II->getIntrinsicID() : Intrinsic::not_intrinsic;
+    switch (ID) {
+    case Intrinsic::assume:
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:990]: probably best to wrap the below below in `{}`

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 9]
```diff
@@ -984,32 +984,37 @@ void State::addInfoFor(BasicBlock &BB) {
...
+      Value *A, *B;
+      CmpInst::Predicate Pred;
+      if (match(I.getOperand(0), m_ICmp(Pred, m_Value(A), m_Value(B)))) {
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:993]: indent level can be reduced by checking `!match()` and have early break

**Review by fhahn — COMMENTED** [review]
Thanks for the latest update, looks mostly good, with a few small remaining comments inline and suggestions for follow-up simplifications.

It looks like the code-formatting checks failed, so it would be good to make sure the code is formatted correctly.

>The test that involves arithmetic was intentionally made simple. Regarding InstSimplify - unfortunately I would probably disagree with the approach. ConstraintElim is a more general / robust solution and iirc it was introduced for these purposes (as an alternative to expanding the complexity of patterns in InstCombine/InstSimplify), but I'd allow @nikic and @fhahn to weigh in / correct me. I think I've seen a few more cases where ConstraintElim stumbles on intrinsics, this was just the simplest one.

I don't think it's a question of either/or, there should be value in handling it in both. InstCombine/InstSimplify is used at more different points in the pipeline (vs ConstraintElimination running once). The main advantage of support in ConstraintElimination isn't simplifying relatively simple stuff like

```
  %abs = tail call i32 @llvm.abs.i32(i32 %arg, i1 false)
   %cmp = icmp sge i32 %abs, %arg
```

but more complicated expressions.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 7 above)
Comment (alexander-shaposhnikov) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1008]: Done

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 5 above)
Comment (alexander-shaposhnikov) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1642]: Done

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 6 above)
Comment (alexander-shaposhnikov) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1645]: yeah

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 8 above)
Comment (alexander-shaposhnikov) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Done

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 9 above)
Comment (alexander-shaposhnikov) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Done

**Comment by alexander-shaposhnikov** [comment]
p.s. clang-format doesn't seem to format "if" properly:
Screenshot 2023-12-28 at 11 35 33 PM

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

> p.s. clang-format doesn't seem to format "if" properly: Screenshot 2023-12-28 at 11 35 33 PM

having multiple statements in an `if` is very uncommon in the codebase I think and should probably be avoided for consistency.

**Comment by fhahn** [comment]
LGTM, thanks!

> p.s. clang-format doesn't seem to format "if" properly: Screenshot 2023-12-28 at 11 35 33 PM

having multiple statements in an `if` is very uncommon in the codebase I think and should probably be avoided for consistency.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 10]
```diff
@@ -1253,7 +1261,7 @@ static std::optional<bool> checkCondition(CmpInst *Cmp, ConstraintInfo &Info,
   Value *B = Cmp->getOperand(1);
 
   auto R = Info.getConstraintForSolving(Pred, A, B);
-  if (R.empty() || !R.isValid(Info)){
+  if (R.empty() || !R.isValid(Info)) {
```
[the remark this answers is not in the document]
Comment (alexander-shaposhnikov) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Removed.

**Review by fhahn — APPROVED** [review]
Still LGTM, thanks!

**Comment by fhahn** [comment]
@alexander-shaposhnikov are you planning on working on the suggestions for follow-ups?

**Comment by alexander-shaposhnikov** [comment]
@fhahn - yup, I'm planning to send a follow-up PR ~this week (if it's not urgent)

**Comment by fhahn** [comment]
> @fhahn - yup, I'm planning to send a follow-up PR ~this week (if it's not urgent)

Sounds great, just wanted to check, it's not urgent at all.

**Pull request merged**